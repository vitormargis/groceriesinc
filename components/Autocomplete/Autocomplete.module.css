.container {
  position: relative;
  margin-top: -150px;
}

.suggestions {
  position: relative;
  border-bottom: 1px solid #ccc;
  border-top-width: 0;
  list-style: none;
  margin-top: 0;
  max-height: 250px;
  overflow-y: auto;
  padding-left: 0;
  width: 300px;
  position: absolute;
  box-shadow: inset 15px 0px 12px rgba(255, 255, 255, 1), 
              inset -15px 0px 12px rgba(255, 255, 255, 1), 
              inset 0px 1px 8px rgba(0, 0, 0, 0.5);
}

.suggestionsItem {
  padding: 15px;
  height: 50px;
  display: flex;
  flex-direction: column;
  justify-content: center;
}

.active {
  background-color: rgb(251, 244, 238);
  cursor: pointer;
  font-weight: 700;
}

.suggestionsItem:not(:last-of-type) {
  border-bottom: 1px solid #ccc;
}

.preInput {
  position: absolute;
  right: 0;
  top: 0;
  pointer-events: none;
  opacity: 0.2;
  font-family: 'Kanit', sans-serif;
  border: none;
  border-bottom: 1px solid #999;
  padding: 20px 15px 15px 45px;
  font-size: 20px;
  width: 300px;
  outline: none; 
}

.input {
  font-family: 'Kanit', sans-serif;
  border: none;
  border-bottom: 1px solid #999;
  padding: 20px 15px 15px 45px;
  font-size: 20px;
  width: 300px;
  outline: none;
  background-image: url('/search-icon.png');
  background-size: 30px;
  background-position: 10px 20px;
  background-repeat: no-repeat;  
}