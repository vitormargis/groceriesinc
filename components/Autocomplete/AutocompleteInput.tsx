import React, { FC, KeyboardEvent, ChangeEvent } from 'react';

interface AutocompleteInputProps {
  input: string,
  preInput: string,
  options: string[],
  onKeyDown: (e: KeyboardEvent<HTMLInputElement>) => void, 
  onChange: (e: ChangeEvent<HTMLInputElement>, options: string[]) => void
}

const AutocompleteInput: FC<AutocompleteInputProps> = ({onChange, onKeyDown, input, preInput, options})  => {
  return <>
    <input
      type="text"
      onChange={(e) => onChange(e, options)}
      onKeyDown={onKeyDown}
      value={input}
      placeholder="Search your fruit"
      className="autocomplete-input"
      data-cy="autocomplete-input"
    />
    <input
      type="text"
      value={preInput}
      className="autocomplete-pre-input"
      data-cy="autocomplete-pre-input"
      readOnly
    />
  </>
}

export default AutocompleteInput; 