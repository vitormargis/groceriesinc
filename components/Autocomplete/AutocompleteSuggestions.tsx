import React, { FC, MouseEvent } from 'react';

interface AutocompleteOptionsProps {
  active: number,
  filtered: string[],
  onMouseEnter: (e: MouseEvent) => void, 
  onClick: (e: MouseEvent) => void
}

const AutocompleteOptions: FC<AutocompleteOptionsProps> = ({active, filtered, onMouseEnter, onClick})  => {
  return filtered.length ? (
    <ul className="autocomplete-suggestions" data-cy="autocomplete-suggestions">
      {filtered.map((suggestion, index) => {
        return (
          <li 
            className={index === active ? 'active' : ''} 
            data-cy={`autocomplete-suggestion-${index}`} 
            key={suggestion} 
            onClick={onClick} 
            onMouseEnter={onMouseEnter}>
            {suggestion}
          </li>
        );
      })}
    </ul>
  ) : null
}

export default AutocompleteOptions; 