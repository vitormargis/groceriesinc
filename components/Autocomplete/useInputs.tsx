import { useState } from "react";

function useInputs(): useInputsInterface {
  const [preInput, setPreInput] = useState('');
  const [searchInput, setSearchInput] = useState('');

  return {
    onPreInput: ({ suggestion, inputValue }) => {
      const isExactMatch = inputValue.toLowerCase() === suggestion.toLowerCase().substring(0, inputValue.length);

      if(inputValue === '' || inputValue === ' ') {
        return setPreInput('')
      }

      if((isExactMatch && suggestion !== '') || suggestion === '') {
        return setPreInput(`${inputValue}${suggestion.substring(inputValue.length, suggestion.length)}`);
      }

      return setPreInput(`${inputValue} ➝ ${suggestion}`);
    },

    onSearchInput: (inputValue) => {
      setSearchInput(inputValue)
    },

    onResetPreInput: (inputValue) => {
      setPreInput('')
      setSearchInput(inputValue)
    },

    searchInput,
    preInput,
  }
}

export default useInputs;

interface useInputsInterface {
  onPreInput: (options: onPreInputInterface) => void,
  onSearchInput: (inputValue: string) => void,
  onResetPreInput: (inputValue: string) => void,
  searchInput: string
  preInput: string
}

interface onPreInputInterface {
  suggestion: string,
  inputValue: string,
}