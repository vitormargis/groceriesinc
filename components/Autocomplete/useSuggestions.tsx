import { useState } from "react";

const useSuggestions = () => {
  const [IndexSuggestion, setIndexSuggestion] = useState(0);
  const [filteredSuggestions, setFilteredSuggestions] = useState([]);

  return {
    onFilterSuggestions: ({options = [], inputValue = ''}) => {
      const inputNormalized = inputValue ? inputValue.toLowerCase() : null;
      
      const filteredMatches = options.filter(option => 
        option.toLowerCase().includes(inputNormalized));
      
      const filteredFirstLetterMatches = filteredMatches.filter(option => 
        option.toLowerCase().startsWith(inputNormalized));

      const cleanFilteredSuggestions = inputNormalized || inputNormalized !== ' '
        ? Array.from(new Set([...filteredFirstLetterMatches, ...filteredMatches])) : []

      setFilteredSuggestions(cleanFilteredSuggestions);
      setIndexSuggestion(0)

      return cleanFilteredSuggestions;
    },

    onPreselectSuggestions: (n = 0) => {
      setIndexSuggestion(IndexSuggestion + (n))
    },

    onResetSuggestions: (options = []) => {
      setFilteredSuggestions(options);
      setIndexSuggestion(0)
    },
    IndexSuggestion,
    filteredSuggestions,
  }
}

export default useSuggestions;