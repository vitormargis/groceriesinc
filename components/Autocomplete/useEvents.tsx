import React, { MouseEvent, KeyboardEvent, ChangeEvent } from "react";
import useSuggestions from './useSuggestions';
import useInputs from './useInputs';

export default function useEvents () {
  const { 
    onPreInput, onSearchInput, onResetPreInput, searchInput, preInput 
  } = useInputs();

  const { 
    onFilterSuggestions, onResetSuggestions, onPreselectSuggestions, filteredSuggestions, IndexSuggestion 
  } = useSuggestions();

  const onArrowNavigation = (e: KeyboardEvent, n: number) => {
    e.preventDefault();
    const suggestion = filteredSuggestions[IndexSuggestion + (n)] || ''
    onPreInput({suggestion, inputValue: searchInput});
    onPreselectSuggestions(n);
  };

  return {
    onChange: (e: ChangeEvent<HTMLInputElement>, options: string[]) => {
        const inputValue = e.currentTarget.value; 
        const suggestion = onFilterSuggestions({options, inputValue})[IndexSuggestion] || '';
        onPreInput({suggestion, inputValue});
        onSearchInput(inputValue)
    },

    onKeyDown: (e: KeyboardEvent) => {
        const suggestion = filteredSuggestions[IndexSuggestion];

        if (e.key === 'Enter') {
          onResetPreInput(suggestion)
          return onResetSuggestions();
        }

        if (e.key === 'ArrowUp') {
          return (IndexSuggestion === 0) ? e.preventDefault() : onArrowNavigation(e, -1)
        }
        
        if (e.key === 'ArrowDown') {
          return (IndexSuggestion + 1 === filteredSuggestions.length) ? e.preventDefault() : onArrowNavigation(e, 1)
        }
    },

    onMouseEnter: (e: MouseEvent<HTMLLIElement>) => {
        const newActive = filteredSuggestions.findIndex(suggestion => suggestion === e.currentTarget.innerText) 
        onPreselectSuggestions(newActive - IndexSuggestion);
        onPreInput({suggestion: filteredSuggestions[newActive], inputValue: searchInput});
    },

    onClick: (e: MouseEvent<HTMLLIElement>) => {
        onResetPreInput(e.currentTarget.innerText)
        onResetSuggestions();
    },

    values: {
      searchInput,
      preInput,
      filteredSuggestions,
      IndexSuggestion
    }
  }
}