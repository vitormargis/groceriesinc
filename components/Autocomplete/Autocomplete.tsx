import React from "react";
import AutocompleteSuggestions from './AutocompleteSuggestions';
import AutocompleteInput from './AutocompleteInput';
import useEvents from './useEvents';

const Autocomplete = ({ options }) => {
  const { onChange,
    onKeyDown,
    onClick,
    onMouseEnter,
    values: {
      searchInput,
      preInput,
      IndexSuggestion,
      filteredSuggestions 
    }
  } = useEvents();

  return (
    <div className="autocomplete-container" data-cy="autocomplete-container">
      <AutocompleteInput 
        onChange={(e) => onChange(e, options)}
        onKeyDown={onKeyDown}
        input={searchInput}
        preInput={preInput}
        options={options}
      />
      <AutocompleteSuggestions
        active={IndexSuggestion}
        filtered={filteredSuggestions}       
        onClick={onClick}
        onMouseEnter={onMouseEnter}
      />
    </div>
  );
}

export default Autocomplete;