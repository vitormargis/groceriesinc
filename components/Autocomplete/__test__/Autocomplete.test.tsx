import React from 'react';
import { act, create } from 'react-test-renderer';
import Autocomplete from '../Autocomplete';

const renderComponent = () => {
  let result: any;
  act(() => {
    result = create(<Autocomplete />);
  });

  return result;
};

describe('Breadcrumbs', () => {
  it('should render with background image and with scrim', () => {
    const component = renderComponent();
    act(() => {
      component.root.findByProps({ preInput: '' }).props.onChange({ currentTarget: { value: '' }})
    });
    expect(component.toJSON()).toMatchSnapshot();
  });
});