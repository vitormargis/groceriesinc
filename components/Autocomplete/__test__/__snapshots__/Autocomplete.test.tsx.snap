// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumbs should render with background image and with scrim 1`] = `
<div
  className="autocomplete-container"
  data-cy="autocomplete-container"
>
  <input
    className="autocomplete-input"
    data-cy="autocomplete-input"
    onChange={[Function]}
    onKeyDown={[Function]}
    placeholder="Search your fruit"
    type="text"
    value=""
  />
  <input
    className="autocomplete-pre-input"
    data-cy="autocomplete-pre-input"
    readOnly={true}
    type="text"
    value=""
  />
</div>
`;
